core/arch/riscv64/riscv64.asm

```diff
@@ -233,7 +230,7 @@ GLOBAL_LABEL(dr_longjmp:)
         /* int atomic_swap(int *adr, int val) */
         DECLARE_FUNC(atomic_swap)
 GLOBAL_LABEL(atomic_swap:)
-/* FIXME i#3544: Not implemented */
+        amoswap.d      ARG1, ARG2, 0 (ARG1)
```

Comment:
We're using the same reg for the source addr and destination. Can that be problematic? Looking at the RISC-V manual on amo operations:

```
load a data value from the address in rs1, place the value into register rd,
apply a binary operator to the loaded value and the original value in rs2, then store the result back
to the address in rs1.
```
My understanding of the above: if rs1 is same as rd, then the loaded value will overwrite the address and the address will be lost, and the later store back into rs1 wouldn't work as expected.

It'll probably be best to try this sequence out in a small asm code.

Reply:
No, it won't be a problem. Also, you can look at [how Spike implemented it](https://github.com/riscv-software-src/riscv-isa-sim/blob/master/riscv/insns/amoswap_d.h) to be sure, for `amoswap`, the address in RS1 is precomputed. The spec tends to not give any pseudo code for reference, so Spike is the golden reference.

Reply:
> We're using the same reg for the source addr and destination. Can that be problematic? Looking at the RISC-V manual on amo operations:
> 
> ```
> load a data value from the address in rs1, place the value into register rd,
> apply a binary operator to the loaded value and the original value in rs2, then store the result back
> to the address in rs1.
> ```
> 
> My understanding of the above: if rs1 is same as rd, then the loaded value will overwrite the address and the address will be lost, and the later store back into rs1 wouldn't work as expected.
> 
> It'll probably be best to try this sequence out in a small asm code.

Some test codes runs in qemu:
```
    #...
    la   a0, print_fmt
    la   t0, data
    li   t2, 12
    amoadd.w t2, t2, 0(t0)
    mv   a1, t2
    call printf
    #...
        la a0, print_fmt
    la t0, data
    addi t2, zero, 1
    amoswap.w t2, t2, 0(t0)
    lw   a1, 0(t0)
    call printf
    #...
```
When I test the code `amoadd.w t0, t2, (t0)`. And then I got a SIGSEGV in ` lw   a1, 0(t0)`. So I think if `amoadd.w t2, t2, (t0)` would be better? Do we really need the value of rd if we only care about the correct operation on *rs1? After all, rd is just a copy of rs1. Isn't that right?

Reply:
Thanks for checking.